### PR TITLE
CDAP-14003: Support outgoing connection lookup for field

### PR DESCRIPTION
**Pull request by rohitsinha54**
### Summary
- Adds support to find outgoing operations for a field.
- When outgoing operation for a field is requested all the `Read` operation that read the field is identified. If there are multiple `Read` operations for that field all of them will be processed. 
- From the originating `Read` operation the operations graph is traversed in depth first order by finding all the outgoing operation from the given read operations. This information is retrieved from the `operationOutgoingConnections `
- If the outgoing operation (either a transform or write) does not have the input as the field for which outgoing operation is required then that path is not explored as it is deduced that the field did not contribute in the operation and hence also did not contributed in any operations following it.
- If the next operation input does contains the field then the operation are recursively explored until a write operation is seen which marks the end of path.
- The outgoing operations are tracked in `outgoingOperations` set across the recursive calls. 

[Issue](https://issues.cask.co/browse/CDAP-14003)
Build: https://builds.cask.co/browse/CDAP-DUT6638

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 1]
```diff
@@ -313,55 +317,63 @@ private long computeChecksum() {
...
+   * the fields which have incoming connection the key field
+   */
+  private void computeIncoming(EndPointField destination, Operation currentOperation,
+                               Operation previousOperation, Set<Operation> incomingOperations,
+                               Map<EndPointField, Set<EndPointField>> result) {
+    if (!incomingOperations.add(currentOperation)) {
       return;
     }
 
+    // TODO (Rohit): This method is doing multiple unrelated things. First this part find outs all the incoming
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: what is there to do? refactor?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 2]
```diff
@@ -313,55 +317,63 @@ private long computeChecksum() {
...
-                                            Map<EndPointField, Set<EndPointField>> result) {
-    if (!visitedOperationNames.add(currentOperation.getName())) {
+  /**
+   * Helper method to compute the incoming connections
+   *
+   * @param destination the {@link EndPointField} whose incoming connection needs to be processed
+   * @param currentOperation the operation being processed (since we are processing incoming this operation is on the
+   * left side if graph is imagined in horizontal orientation or this operation is the input to the to
+   * previousOperation
+   * @param previousOperation the previous operation which is processed and reside on right to the current operation if
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I know this was the existing name, but it seems more clear to name this 'nextOperation'. It's confusing that 'previous' is the opposite of what one would normally expect.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 3]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+  private Set<InputField> getInputFields(Operation nextOperation) {
+    Set<InputField> inputFields;
+    if (OperationType.WRITE == nextOperation.getType()) {
+      WriteOperation nextWrite = (WriteOperation) nextOperation;
+      inputFields = new HashSet<>(nextWrite.getInputs());
+    } else if (OperationType.TRANSFORM == nextOperation.getType()) {
+      TransformOperation nextTransform = (TransformOperation) nextOperation;
+      inputFields = new HashSet<>(nextTransform.getInputs());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either only any of the following '%s'operations as " +
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: this should never happen right? Is it validated somewhere?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 4]
```diff
@@ -313,55 +317,63 @@ private long computeChecksum() {
...
+   * left side if graph is imagined in horizontal orientation or this operation is the input to the to
+   * previousOperation
+   * @param previousOperation the previous operation which is processed and reside on right to the current operation if
+   * the graph is imagined to be in horizontal orientation.
+   * @param incomingOperations all the operations which are processed so far
+   * @param result a {@link Map} of {@link EndPointField} to {@link Set} of {@link EndPointField} which represents all
+   * the fields which have incoming connection the key field
+   */
+  private void computeIncoming(EndPointField destination, Operation currentOperation,
+                               Operation previousOperation, Set<Operation> incomingOperations,
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I think the previous name was better. 'incomingOperations' sounds like the operations that are input to the current operation.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 5]
```diff
@@ -313,55 +317,63 @@ private long computeChecksum() {
...
-      Set<InputField> inputFields = new HashSet<>();
-      if (OperationType.WRITE == previousOperation.getType()) {
-        WriteOperation previousWrite = (WriteOperation) previousOperation;
-        inputFields = new HashSet<>(previousWrite.getInputs());
-      } else if (OperationType.TRANSFORM == previousOperation.getType()) {
-        TransformOperation previousTransform = (TransformOperation) previousOperation;
-        inputFields = new HashSet<>(previousTransform.getInputs());
-      }
 
+      Set<InputField> inputFields = getInputFields(previousOperation);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: In this case, I think it's better not to have this additional method. It splits up logic that is more clear together. For example, when looking at the getInputFields method, it was not clear to me why OperationType.READ was invalid, but here it is clear, as we don't expect to have a READ connected to another READ.

Comment (albertshau) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: If anything, I would have a method to handle when the current operation is a read and a method to handle when it is a transform. Everything in the if clause could move into that method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 6]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
   }
 
   /**
-   * Get the subset of operations that were responsible for computing the specified field of
-   * a specified destination.
+   * <p>Get the subset of operations that were responsible for computing the specified field of
+   * a specified destination.</p>
+   * <p>For example if the operation are as follow</p>
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java:394]: good example to clarify the javadoc

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 7]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+   * <pre>
+   * pRead: personFile -> (offset, body)
+   * parse: body -> (id, name, address)
+   * cRead: codeFile -> id
+   * codeGen: (parse.id, cRead.id) -> id
+   * sWrite: (codeGen.id, parse.name, parse.address) -> secureStore
+   * iWrite: (parse.id, parse.name, parse.address) -> insecureStore
+   * </pre>
+   * <p>The destination field is 'id' field of insecureStore then the result set will contain the operations iWrite,
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: The -> If the

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 8]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+   * <pre>
+   * pRead: personFile -> (offset, body)
+   * parse: body -> (id, name, address)
+   * cRead: codeFile -> id
+   * codeGen: (parse.id, cRead.id) -> id
+   * sWrite: (codeGen.id, parse.name, parse.address) -> secureStore
+   * iWrite: (parse.id, parse.name, parse.address) -> insecureStore
+   * </pre>
+   *
+   * <p>The sourceField  is 'id' of codeFile then the returned operations set will contain cRead, codeGen, sWrite.</p>
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: extra space 'sourceField  is'

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 9]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+   * sWrite: (codeGen.id, parse.name, parse.address) -> secureStore
+   * iWrite: (parse.id, parse.name, parse.address) -> insecureStore
+   * </pre>
+   * <p>The destination field is 'id' field of insecureStore then the result set will contain the operations iWrite,
+   * parse, pRead.</p>
+   * <p>If the destination field is 'id' field of secureStore then the result set will contain the operations sWrite,
+   * codeGen, parse, pRead, cRead.</p>
    *
    * @param destinationField the EndPointField for which the operations need to find out
+   *
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: nit: for consistency, don't separate with a newline

Comment (albertshau) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: same for other javadocs in this class.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 10]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+        List<String> inputs = readOperation.getOutputs();
+        for (String input : inputs) {
+          if (input.equals(sourceField.getField())) {
+            // if the read operation read the sourceField
+            // add the read operation to outgoingOperations
+            outgoingOperations.add(readOperation);
+            // find all the outgoing operation from this read operations and if they are connected i.e. the next
+            // operation did used the field for which the outgoing operations is requested then traverse that
+            // operation further
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: this much nesting makes things difficult to understand. I think it would be cleaner to do something like:

```
ReadOperation readOperation = null;
for (ReadOperation operation : readOperations) {
  if (operation.getSource().equals(sourceField.getEndPoint()) {
    operation = readOperation;
    break;
  }
}

boolean fieldIsOutput = false;
for (String output : readOperation.getOutputs()) {
  if (output.equals(sourceField.getField())) {
    fieldIsOutput = true;
    break;
  }
}

if (fieldIsOutput) {
  ...
}
```

This would also prevent unnecessary looping that is currently happening.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 11]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+  Set<Operation> getOutgoingOperationsForField(EndPointField sourceField) {
+    if (readOperations == null) {
+      computeAndValidateFieldLineageInfo(this.operations);
+    }
+
+    Set<Operation> outgoingOperations = new HashSet<>();
+    for (ReadOperation readOperation : readOperations) {
+      if (readOperation.getSource().equals(sourceField.getEndPoint())) {
+        // if the read operation is for the dataset to which the sourceField belong
+        List<String> inputs = readOperation.getOutputs();
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: seems like this is misnamed, shouldn't it be outputs?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 12]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+        for (String input : inputs) {
+          if (input.equals(sourceField.getField())) {
+            // if the read operation read the sourceField
+            // add the read operation to outgoingOperations
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: this comment is confusing to me. Isn't this adding the ReadOperation for the endpoint itself as an 'outgoing' operation?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 13]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+            outgoingOperations.add(readOperation);
+            // find all the outgoing operation from this read operations and if they are connected i.e. the next
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: this comment is confusing. Can you add an example of what is happening here from the example you gave in the javadoc for the method?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 14]
```diff
@@ -382,41 +394,174 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+            // operation did used the field for which the outgoing operations is requested then traverse that
+            // operation further
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
+              if (isUsed(readOperation, outgoingOperation, sourceField.getField())) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: why is this needed? Is it because we don't want 'outgoingOperation' to be included in the result because it's a dead end? But isUsed doesn't seem to be recursive, so doesn't that mean dead ends can be included anyway?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Moved into separate methods to make them do one thing.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: This the the next operation in DAG although it is the previous operation in the traversal order as we are traversing backward. 

Should we name it nextOperation then? I agree previousOperation is confusing but nextOperation might be confusing too. Thoughts?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 5 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done. Moved it back as it seem to give better flow.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 2 above)
Comment (albertshau) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: hm I see. Let's keep it as is.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 7 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 9 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 8 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 11 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Done.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 12 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Cleared.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 14 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I added a comment here in code for clarification. Please let me know if it is helpful.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 10 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: We want to do this for all readOperations which reads from the source. If R1 and R2 both read the sourceField then we want both to be processed and included in result set being returned.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 13 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Removed that comment and added a detailed comment below it. Let me know if it is still hard to understand.

**Review by rohitsinha54 — COMMENTED** [review]
@albertshau:

Address comments. 
1. Commit 1: Reviewed

For review:
2. [Commit 2](https://github.com/caskdata/cdap/pull/10654/commits/daeef2a5282a540b1f06810307510a78d877ed19): Incoming operation cleanup
3. [Commit 3](https://github.com/caskdata/cdap/pull/10654/commits/e0503e08ab84745e3861fb23982df9be753f8182): Addressed comments from review.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 15]
```diff
@@ -309,26 +311,32 @@ private long computeChecksum() {
...
-                                            Map<EndPointField, Set<EndPointField>> result) {
-    if (!visitedOperationNames.add(currentOperation.getName())) {
-      return;
-    }
-
+  /**
+   * Helper method to compute the incoming summary
+   *
+   * @param field the {@link EndPointField} whose summary needs to be calculated
+   * @param currentOperation the operation being processed (since we are processing incoming this operation is on the
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: processed. Since we are

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 16]
```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
+              // check if the outgoing operation does use the sourceField else that even though is an outgoing
+              // operation for the current operation (which reads multiple fields) is not an outgoing operation for
+              // this field as it does not uses it. For example in the above example included in the java doc of this
+              // method the 'parse' transform is an outgoing operation for the 'pread' operation and is also an
+              // outgoing operation for the 'body' field of 'personFile' dataset read by 'pRead' operation but
+              // 'parse' is not an outgoing operation for the field 'offset' of 'personFile' as it does not takes it
+              // as input.
+              if (isUsed(readOperation, outgoingOperation, sourceField.getField())) {
+                computeOutgoing(outgoingOperation, visitedOperations);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I still think there is too much nesting here, even though the suggested implementation was invalid. This line is in the middle of a for - if - for - if - for - if.

You can remove 3 layers by checking for the opposite and continuing:
```
if (!readOperation.getSource().equals(sourceField.getEndpoint()) {
  continue;
}
List<String> outputs ...
```

That still leaves a for inside a for inside a for, but it should improve readability.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 17]
```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+      if (readOperation.getSource().equals(sourceField.getEndPoint())) {
+        // if the read operation is for the dataset to which the sourceField belong
+        List<String> outputs = readOperation.getOutputs();
+        for (String output : outputs) {
+          // if the current read operation did read the sourceField for which outgoing operation is requested we
+          // consider it else not.
+          if (output.equals(sourceField.getField())) {
+            visitedOperations.add(readOperation);
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
+              // check if the outgoing operation does use the sourceField else that even though is an outgoing
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I think it would be more clear to do say something like:

```
Check that the source field is an input field for the outgoing operation.
Consider the example in the method javadoc with:
  sourceField = personFile.offset
  readOperation = pRead: personFile -> (offset, body)
  outgoingOperation = parse: body -> (id, name, address)

In this scenario, 'offset' is not an input to the outgoingOperation so we do not need to go further down the graph.
If the sourceField was personFile.body, we would need to continue as 'body' is an input to the outgoingOperation.
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 18]
```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+   * Checks whether the given fieldName is used in the next operations or not
+   *
+   * @param previousOperation previous {@link ReadOperation}
+   * @param nextOperation the next operation which should either be a {@link TransformOperation} or {@link
+   * WriteOperation}
+   * @param fieldName the name of the field whose usage needs to be checked
+   * @return true if the field is used in the nextOperation
+   */
+  private boolean isUsed(ReadOperation previousOperation, Operation nextOperation, String fieldName) {
+    if (!previousOperation.getOutputs().contains(fieldName)) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: try to avoid creating methods that allow the possibility of passing in invalid input. As suggested above, I think it would be better to make this:

```
private boolean containsInputField(Operation operation, InputField inputField) {
  ...
}
```
Then you never need to worry about invalid arguments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 19]
```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+    Set<InputField> inputFields = new HashSet<>();
+    if (OperationType.WRITE == nextOperation.getType()) {
+      WriteOperation nextWrite = (WriteOperation) nextOperation;
+      inputFields = new HashSet<>(nextWrite.getInputs());
+    } else if (OperationType.TRANSFORM == nextOperation.getType()) {
+      TransformOperation nextTransform = (TransformOperation) nextOperation;
+      inputFields = new HashSet<>(nextTransform.getInputs());
+    }
+    // if the next operation inputFields does contains the given fieldName return true
+    for (InputField inputField : inputFields) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: can be simplified to `return inputFields.contains(inputField)`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java [hunk 20]
```diff
@@ -382,41 +389,188 @@ private void computeIncomingSummaryHelper(EndPointField destination, Operation c
...
+            visitedOperations.add(readOperation);
+            for (Operation outgoingOperation : operationOutgoingConnections.get(readOperation.getName())) {
+              // check if the outgoing operation does use the sourceField else that even though is an outgoing
+              // operation for the current operation (which reads multiple fields) is not an outgoing operation for
+              // this field as it does not uses it. For example in the above example included in the java doc of this
+              // method the 'parse' transform is an outgoing operation for the 'pread' operation and is also an
+              // outgoing operation for the 'body' field of 'personFile' dataset read by 'pRead' operation but
+              // 'parse' is not an outgoing operation for the field 'offset' of 'personFile' as it does not takes it
+              // as input.
+              if (isUsed(readOperation, outgoingOperation, sourceField.getField())) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I think it would be more clear to do something like:

```
InputField inputToOutgoingOperation = new InputField(sourceField.getEndPoint().getName(), sourceField.getField());
if (containsInputField(outgoingOperation, inputToOutgoingOperation)) {
  computeOutgoing(...);
}
```

By the way, it seems like this class ignores the namespace of the endpoint everywhere and just uses the name. Isn't that problematic?

**Review by albertshau — COMMENTED** [review]
small comments

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 19 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: Used: 
```
return inputFields.contains(InputField.of(previousOperation.getName(), fieldName));
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 18 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: 💯 
Thanks.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java (hunk 20 above)
Comment (rohitsinha54) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java]: I think the class names are little confusing to keep track in head while reading the code. Namespace is only contained in EndPoint which is also a member of EndPointField. We don't do .getName() on EndPoint in this class. 
The confusion might be because of the .getName() of InputField which the name returns the field name and the origin is the Operation name which was the origin of this input field

**Review by rohitsinha54 — COMMENTED** [review]
@albertshau Addressed comments. Please have a look.

**Review by albertshau — DISMISSED** [review]
lgtm

**Review by gsps1 — APPROVED** [review]
